### PR TITLE
LibJS: The Object Rewrite

### PR DESCRIPTION
**Pull request by linusg**
~Currently a bunch of things are still broken.~ :white_check_mark:
~I'll split it into multiple commits (or PRs) once that's sorted out :^)~ :white_check_mark:

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/Value.h [hunk 1]
```diff
@@ -301,6 +301,8 @@ class Value {
 
     String typeof() const;
 
+    bool operator==(Value const) const;
```
Comment (awesomekling) [inline Userland/Libraries/LibJS/Runtime/Value.h:304]: `Value const&` perhaps?

Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/Value.h]: Fixed :+1:

File: Userland/Libraries/LibJS/Runtime/Array.cpp [hunk 2]
```diff
@@ -23,16 +23,27 @@ Array* Array::create(GlobalObject& global_object, size_t length, Object* prototy
...
     return array;
 }
 
 // 7.3.17 CreateArrayFromList ( elements ), https://tc39.es/ecma262/#sec-createarrayfromlist
-Array* Array::create_from(GlobalObject& global_object, const Vector<Value>& elements)
+Array* Array::create_from(GlobalObject& global_object, Vector<Value> const& elements)
 {
+    // 1. Assert: elements is a List whose elements are all ECMAScript language values.
+
+    // 2. Let array be ! ArrayCreate(0).
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/Array.cpp:35]: Not that it's a bad thing, but any reason for the very verbose spec copy here?

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.h [hunk 3]
```diff
@@ -102,7 +102,7 @@ static Value raw_bytes_to_numeric(GlobalObject& global_object, ByteBuffer raw_va
     }
 }
 
-// 25.1.2.10 GetValueFromBuffer ( arrayBuffer, byteIndex, type, isTypedArray, order [ , isLittleEndian ] ), https://tc39.es/ecma262/#sec-getvaluefrombuffer
+// Implementation for 25.1.2.10 GetValueFromBuffer, used in TypedArray<T>::get_value_from_buffer().
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.h:105]: Odd change, why remove the arguments+link and mention where it's used instead?

File: Userland/Libraries/LibJS/Runtime/ArrayConstructor.cpp [hunk 4]
```diff
@@ -139,7 +139,7 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayConstructor::from)
                 return {};
 
             if (!next) {
-                array_object.put(vm.names.length, Value(k));
+                array_object.set(vm.names.length, Value(k), true);
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/ArrayConstructor.cpp:142]: I honestly have no idea what `true` is for, can we have an enum class instead?

**Review by alimpfard — COMMENTED** [review]
Overall, not the biggest fan of "copy all the spec steps into the code" change, especially when they're painfully obvious
e.g. 
```
// Else
else {
```
But I'll respect that if it's an explicit decision on your part.

**Review by awesomekling — APPROVED** [review]
Given the huge amount of test coverage, and the nature of these changes, I think we can go ahead and merge them without thinking too much. :^)

As you say, we have a bunch of performance work ahead of us, but this is a much stronger foundation to do that work off of. Excellent work!

File: Userland/Libraries/LibJS/Runtime/ArrayConstructor.cpp (hunk 4 above)
Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/ArrayConstructor.cpp:142]: no, the specification requires that parameter to be a boolean

File: Userland/Libraries/LibJS/Runtime/Array.cpp (hunk 2 above)
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/Array.cpp:35]: Multiple reasons:

- you can pretty much immediately see if a function is fully or partially spec compliant - if it has these comments everywhere I'm trusting it, and if it doesn't I'm much more sceptical if it really does what it should do
- it encourages spec compliant implementations of new code (similar to adding AO names and links - it's much harder to half-ass something if a comment one line above tells a different story)
- it's easier to see when the code gets out of sync with the spec, which it likely will at some point - the spec evolves (other engines simply use `// 1.`, `// 2.` for that, but if we already add comments we might as well add the text, too)
- easier to review, maybe?

It's quite verbose, I know, and I'm not expecting anyone else to do this (kling already did) - but removing it is not up for debate :)

File: Userland/Libraries/LibJS/Runtime/ArrayConstructor.cpp (hunk 4 above)
Comment (alimpfard) [reply Userland/Libraries/LibJS/Runtime/ArrayConstructor.cpp:142]: The spec also requires the "hint" argument to `get_iterator` to be a string, I don't see your point.

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.h (hunk 3 above)
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/ArrayBuffer.h:105]: We had two functions with the `GetValueFromBuffer` comment on them, `get_value_from_buffer()` is the right one - it just happens to wrap this.

File: Userland/Libraries/LibJS/Runtime/ArrayConstructor.cpp (hunk 4 above)
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/ArrayConstructor.cpp:142]: that's true, it's indeed a 1:1 spec translation, which doesn't always make sense - not opposed to change it to a yes/no enum. `Object::SetThrow::Yes/No`?

Comment (IdanHo) [reply Userland/Libraries/LibJS/Runtime/ArrayConstructor.cpp:142]: i mean, thats not true, the hint in getIterator is a type, not a string, but see the part in parenthesis

File: Userland/Libraries/LibJS/Runtime/Array.cpp (hunk 2 above)
Comment (alimpfard) [reply Userland/Libraries/LibJS/Runtime/Array.cpp:35]: Alright, if that's something you want to do, then go for it.

**Review by alimpfard — APPROVED** [review]

File: Userland/Libraries/LibJS/Runtime/Object.cpp [hunk 5]
```diff
@@ -113,881 +68,1009 @@ Object::~Object()
...
-        m_shape = shape.create_prototype_transition(new_prototype);
-    return true;
+
+    // 6. Return success.
+    return success;
 }
 
-bool Object::has_prototype(const Object* prototype) const
+// 7.3.5 CreateDataProperty ( O, P, V ), https://tc39.es/ecma262/#sec-createdataproperty
+bool Object::create_data_property(PropertyName const& property_name, Value value)
```
Comment (mattco98) [inline Userland/Libraries/LibJS/Runtime/Object.cpp:123]: This seems to be the only method without a `VERIFY(property_name.is_valid());` at the beginning

File: Userland/Libraries/LibJS/Runtime/ArrayConstructor.cpp (hunk 4 above)
Comment (alimpfard) [reply Userland/Libraries/LibJS/Runtime/ArrayConstructor.cpp:142]: > Object::SetThrow::Yes/No

That's what I meant by enum class to begin with

Comment (alimpfard) [reply Userland/Libraries/LibJS/Runtime/ArrayConstructor.cpp:142]: Let's worry about this later, this PR is all copy-the-spec anyway, changing one bool isn't gonna suddenly make anything better or worse

File: Userland/Libraries/LibJS/Runtime/Object.cpp (hunk 5 above)
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/Object.cpp:125]: Fixed, thanks!

**Pull request merged**